lib/instrumentation/span.js

```diff
@@ -248,6 +248,21 @@ Span.prototype._encode = function (cb) {
     }
   }
 
+  if (self.isComposite()) {
+    payload.composite = self._compression.encode()
+    payload.timestamp = self._compression.timestamp
+    payload.duration = self._compression.duration
+  }
+
+  if (self.isCompositeSameKind()) {
+    // rename
+    let callsTo = self && self._destination && self._destination.service && self._destination.service.resource
+    if (!callsTo) {
+      callsTo = 'unknown'
+    }
+    payload.name = `Calls to ${callsTo}`
+  }
```

Comment:
This results in this incorrect span name for the first span:

    trace 8666e4
    `- transaction 37e147 "t0" (30.227ms, outcome=unknown)
       `- span ca9816 "Calls to mysql" (11.748ms, mysql)
       `- span e21dc6 "s2" (11.564ms, mysql

using this example code:

```js
var apm = require('./').start({
  metricsInterval: '0s',
  centralConfig: false,
  cloudProvider: 'none',
  spanCompressionEnabled: true
})

// Create two sibling mysql spans that *would* be compressed with "same_kind"
// but don't because they are shorter than the 50ms default threshold.
function main () {
  var t0 = apm.startTransaction('t0')
  setImmediate(() => {
    var s1 = t0.startSpan('s1', 'db', 'mysql')
    s1.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
    setTimeout(() => {
      s1.end()
      var s2 = t0.startSpan('s2', 'db', 'mysql')
      s2.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
      setTimeout(() => {
        s2.end()
        t0.end()
      }, 10)
    }, 10)
  })
}

main()
```

Is the "Calls to ..." naming in `span._encode()` necessary given that naming is also being done in `SpanCompression.prototype.tryToCompressRegular`? If I comment that block out, then the above example no longer fails, and the tests pass.

Reply:
That's definitely some old/dead code in the `_encode` method, so I've removed that. 

However I'm not quite sure I understood your example and wasn't able to get it to fail in the way you indicated (at least with the feature as it was Monday).  

Here's what I mean -- in the code above I see two spans, each of which should be around 10ms long.

The spans have different names but otherwise look similar, so if compression is enabled they should be compressed as same kind.

Each of the spans is less than 10ms which means they _should_ be compressed into a single span with the _Calls to_ name since 10ms is less than the 50ms default `spanCompressionSameKindMaxDuration`. 

The current code appears to do this correctly. I've adapted your code into a test named _create two sibling mysql spans ..._ that demonstrates this.

I was not able to reproduce the behavior you describe where you end up with a tree that looks like this.

```
    trace 8666e4
    `- transaction 37e147 "t0" (30.227ms, outcome=unknown)
       `- span ca9816 "Calls to mysql" (11.748ms, mysql)
       `- span e21dc6 "s2" (11.564ms, mysql
```

Also I don't understand quite understand this comment (unless it's describing undesirable behavior you saw last week?)

```
// Create two sibling mysql spans that *would* be compressed with "same_kind"
// but don't because they are shorter than the 50ms default threshold.
```      

So -- are you still seeing problems with the above code with the up-to-date branch?  Or have I. made some dreadful presumption about something in the spec that's incorrect?

Reply:
First, sorry: I botched that *comment* and made this all confusing.  Let me try again.

I'm running this example with commit 8979b453be7b893150ca9a9975cf51b5f39134c6 (i.e. before your https://github.com/elastic/apm-agent-nodejs/pull/2623/commits/06c06efd1eebbd3e691b16b1a372d0e619b0d21a to remove the defunct code in Span.prototype._encode).

The new code example is as follows. I've only updated the comments from the above code example:

```js
var apm = require('./').start({
  metricsInterval: '0s',
  centralConfig: false,
  cloudProvider: 'none',
  spanCompressionEnabled: true
})

function main () {
  var t0 = apm.startTransaction('t0')
  setImmediate(() => {
    var s1 = t0.startSpan('s1', 'db', 'mysql')
    s1.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
    setTimeout(() => {
      // `s1` is compression eligible, so we expect this to get buffered.
      s1.end()

      var s2 = t0.startSpan('s2', 'db', 'mysql')
      s2.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
      setTimeout(() => {
        // `s2` is a `same_kind` match to s1. However,
        // `spanCompressionSameKindMaxDuration` defaults to 0ms and this is,
        // obviously, longer than that. So we expect *no compression*.
        s2.end()
        t0.end()
      }, 10)
    }, 10)
  })
}

main()
```

My expection is a transaction `t0` with separate spans `s1` and `s2` -- i.e. **no compression ** -- because they are `same_kind` match, but have duration ~10ms which is longer than the default `spanCompressionSameKindMaxDuration: '0ms'`.

What I observe is:

```
    {
        "span": {
            "name": "Calls to mysql",
            "type": "db",
            "id": "ece21bbe8bc46317",
            "transaction_id": "650eb8284d6c4831",
            "parent_id": "650eb8284d6c4831",
            "trace_id": "363fd5101e87cbac9ce72b3a29be63b9",
            "subtype": "mysql",
            "action": null,
            "timestamp": 1649183613812077,
            "duration": 11.721,
            "context": {
                "destination": {
                    "service": {
                        "name": "mysql",
                        "resource": "mysql",
                        "type": "db"
                    }
                }
            },
            "sync": false,
            "outcome": "success",
            "sample_rate": 1
        }
    }
    {
        "span": {
            "name": "s2",
            "type": "db",
            "id": "dd9eeda8bdbf45a4",
            "transaction_id": "650eb8284d6c4831",
            "parent_id": "650eb8284d6c4831",
            "trace_id": "363fd5101e87cbac9ce72b3a29be63b9",
            "subtype": "mysql",
            "action": null,
            "timestamp": 1649183613824160,
            "duration": 11.388,
            "context": {
                "destination": {
                    "service": {
                        "name": "mysql",
                        "resource": "mysql",
                        "type": "db"
                    }
                }
            },
            "sync": false,
            "outcome": "success",
            "sample_rate": 1
        }
    }
```

- Good: The spans were not compressed.
- Bad: The name of the first span was incorrectly changed to `Calls to mysql`.

If I update to the latest commit, then this is fixed and I observe what is expected. In other words, all good now with the latest changes.